designs/sam-debugging/local-sam-debugging.md

```diff
@@ -0,0 +1,338 @@
+
+The Toolkit implements a Debug Configuration type `aws-sam`. Users can author and maintain these configuration entries, then launch them by pressing F5 (or Ctrl+F5 to Run without Debugging). When launched, this configuration type:
+
+-   validates debug configuration inputs (see [Debug Configuration Validations](#debug-configuration-validation))
+-   uses SAM CLI to build a SAM Application
+-   uses SAM CLI to invoke a SAM Template resource
+-   attaches a debugger to the SAM invocation (skipped if "Run without Debugging" was used)
+    -   if a debugger is attached, debug output is shown in the Debug Console
+-   throughout this process, SAM CLI output is shown in the Toolkit's Output Channel
+
+These debug configurations are authored in a json file. The following Toolkit assistance is provided:
+
+-   autocompletion with descriptions is provided for `aws-sam` related fields
+    -   There is no autocompletion available for specific values in a configuration. For example, if a user types in the location of a SAM Template file, there is no filesystem-based autocompletion. The Debug Configuration validates the configuration and notifies of errant values when it is run.
+-   snippets to produce typical (or starter) `aws-sam` debug configurations
+-   when no launch.json file is present in a workspace, VS Code exposes functionality that allows users to request auto-generated Debug Configurations. In this situation, the toolkit generates an `aws-sam` Debug Configuration for all `AWS::Serverless::Function` resources detected within all SAM Templates located in the workspace.
+
+Debug Configurations support local debugging for both [SAM Template resources](#sam-template-resource-local) and [standalone Lambda handlers](#standalone-lambda). Support for these different features is provided using different `request` field values in an `aws-sam` debug configuration.
+
+SAM Template resources are configured by setting `request` to `template-invoke`. A SAM Template and resource are required. This debug configuration uses the SAM Template as-is to debug as discussed above.
+
+Standalone Lambda handlers are configured by setting `request` to `standalone-lambda`. At a minimum, these configurations need to know what the Lambda handler is, which path is considered the project root, and what the Lambda runtime is. When this configuration is launched, a temporary SAM Template is produced. The template has a resource which references the Lambda handler. This temporary template (and resource) is then handled the same way as `template-invoke` configurations are.
+
+Example Debug Configuration entries can be found in the [Appendix](#sample-debug-configurations).
+
+### <a id="codelenses"></a> CodeLenses
+
+#### CodeLenses in SAM Template files
+
+CodeLenses are added to SAM Template files that serve as shortcuts to `template-invoke` [debug configurations](#debug-configurations) defined in the workspace.
```

Comment:
I wonder how useful this will be, developers will be in the `.js` file while they're developing and want to debug their code.

Reply:
Its a valid point. I suspect you'd only use these if you were looking at your application from a high level, and using the template as the starting point. The JetBrains toolkit provides launch functionality from the template files as well.